instrumentation-tests/src/androidTest/java/com/mapbox/navigation/instrumentation_tests/core/TripSessionsBillingTest.kt

```diff
@@ -0,0 +1,316 @@
+import com.mapbox.navigation.instrumentation_tests.utils.http.HttpServiceEventsObserver
+import com.mapbox.navigation.instrumentation_tests.utils.parameters
+import com.mapbox.navigation.instrumentation_tests.utils.routes.RoutesProvider
+import com.mapbox.navigation.instrumentation_tests.utils.routes.RoutesProvider.toNavigationRoutes
+import com.mapbox.navigation.testing.ui.BaseCoreNoCleanUpTest
+import com.mapbox.navigation.testing.ui.utils.getMapboxAccessTokenFromResources
+import com.mapbox.navigation.testing.ui.utils.runOnMainSync
+import kotlinx.coroutines.delay
+import kotlinx.coroutines.flow.Flow
+import kotlinx.coroutines.flow.filter
+import kotlinx.coroutines.flow.take
+import org.junit.After
+import org.junit.Assert.assertEquals
+import org.junit.Assert.assertNotNull
+import org.junit.Assert.assertTrue
+import org.junit.Assert.fail
+import org.junit.Before
+import org.junit.Rule
+import org.junit.Test
+import java.net.URI
+import java.util.concurrent.TimeUnit
+
+class TripSessionsBillingTest : BaseCoreNoCleanUpTest() {
+
+    @get:Rule
+    val mapboxNavigationRule = MapboxNavigationRule()
+
+    private lateinit var mapboxNavigation: MapboxNavigation
+
+    private val route = RoutesProvider.dc_very_short(context).toNavigationRoutes()[0]
+
+    private lateinit var httpEventsObserver: HttpServiceEventsObserver
+
+    override fun setupMockLocation(): Location = mockLocationUpdatesRule.generateLocationUpdate {
+        route.routeOptions.coordinatesList().first().run {
+            latitude = latitude()
+            longitude = longitude()
+        }
+        bearing = 190f
+    }
+
+    @Before
+    fun setup() {
+        httpEventsObserver = HttpServiceEventsObserver.register()
+
+        runOnMainSync {
+            mapboxNavigation = MapboxNavigationProvider.create(
+                NavigationOptions.Builder(context)
+                    .accessToken(getMapboxAccessTokenFromResources(context))
+                    .routingTilesOptions(
+                        RoutingTilesOptions.Builder()
+                            .tilesBaseUri(URI(mockWebServerRule.baseUrl))
+                            .build()
+                    )
+                    .build()
+            )
+        }
+    }
+
+    @After
+    fun cleanUp() {
+        HttpServiceFactory.getInstance().setInterceptor(null)
+    }
+
+    /**
+     * Waits for exactly [count] elements that match [predicate] to appear
+     * in [HttpServiceEventsObserver] to avoid tests flakiness,
+     * than waits for 1 second for more http events to arrive.
+     * This is made in order to catch bugs
+     * where more than expected billing events sent to the server.
+     *
+     * @return All billing [HttpServiceEvent.Request]'s appeared within waiting time.
+     * Number of returned items can be more that [count].
+     * Returned elements might not match the [predicate].
+     */
+    private suspend fun waitForBillingEvents(
```

Comment:
Thanks for writing the doc. Some points were not obvious.
How about the name `getAllBillingEventsWhenHasEvents` for the method? IMO it would make it more clear even without the doc. I know it's longer but it would make it easier to remember what it does when reusing it.
Anyway, up to you.

Reply:
ok, we can use that name.